File: nodejs/Puzzle.js
module.exports = class Puzzle {
    constructor(left, top, right, bottom) {
        this.edges = [left, top, right, bottom];
        this.origin = [left, top, right, bottom];
        this.active = true;
        this.rotation = 0;
        this.atFirstPosition = 0;
    }

    get left() {
        return this.edges[0];
    }

    get top() {
        return this.edges[1];
    }

    get right() {
        return this.edges[2];
    }

    get bottom() {
        return this.edges[3];
    }

    Rotate() {
        const firstEdge = this.edges.shift();
        this.edges.push(firstEdge);
        this.rotation++;
    }

    Reset() {
        this.edges = [...this.origin];
        this.active = true;
        this.rotation = 0;
    }
}

File: go/puzzle.go
package main

type Puzzle struct {
	Origin        []int
	Edges         []int
	Rotation      int
	AtFirstSquare int
	Active        bool
}

func (puzzle *Puzzle) Left() int {
	return puzzle.Edges[0]
}

func (puzzle *Puzzle) Top() int {
	return puzzle.Edges[1]
}

func (puzzle *Puzzle) Right() int {
	return puzzle.Edges[2]
}

func (puzzle *Puzzle) Bottom() int {
	return puzzle.Edges[3]
}

func (puzzle *Puzzle) Rotate() {
	puzzle.Edges = append(puzzle.Edges[1:], puzzle.Edges[0])
	puzzle.Rotation ++
}

func (puzzle *Puzzle) Reset() {
	copy(puzzle.Edges, puzzle.Origin)
	puzzle.Active = true
	puzzle.Rotation = 0
}

func NewPuzzle(left, top, right, bottom int) Puzzle {
	return Puzzle{
		Origin:        []int{left, top, right, bottom},
		Edges:         []int{left, top, right, bottom},
		Rotation:      0,
		AtFirstSquare: 0,
		Active:        true,
	}
}
